### PR TITLE
Add preflight check implementation, add permission validation preflight check

### LINKED ISSUE #855: Optional ability to check if client has permissions to perform CRUD actions before trying them
**Describe the problem/challenge you have**
When deploying an app, if/when kapp encounters an error, it stops immediately. This can leave the content on the cluster in some intermediate or unknown state.

**Describe the solution you'd like**
Add an optional setting for kapp to check if the client has permissions to perform every operation needed before executing any of them. If any of the checks fails, kapp makes no modifications to the cluster and returns an error.

**Anything else you would like to add:**
The SubjectAccessReview API can be used for these checks.

The related kapp-controller issue is https://github.com/carvel-dev/kapp-controller/issues/1381.

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

### PR DESCRIPTION
**Pull request by everettraven**
#### What this PR does / why we need it:

- Adds a flag (`--preflight`) to `kapp deploy` that can be used to enable/disable preflight checks. Can be used in the format `--preflight=CheckName=true,...`
- Adds a preflight check for permission validation that is disabled by default. Can be enabled with `--preflight=PermissionValidation=true`
    - For `(Cluster)Roles` it will ensure that a user has all the necessary permissions to create/update the resource without privilege escalation _or_ has the "escalate" permissions
    - For `(Cluster)RoleBindings` it will ensure that a user has all the necessary permissions to create/update the resource without privilege escalation _or_ has the "bind" permissions
    - For all other resources and operations, it will ensure it has the necessary permissions to perform the changes required by the change set

#### Which issue(s) this PR fixes:

Fixes #855 

#### Does this PR introduce a user-facing change?

```release-note
Adds `--preflight` flag to the `kapp deploy` subcommand. This flag can be used to enable/disable specific preflight checks and follows the format `--preflight=CheckName=true,...`

Adds a PermissionValidation preflight check, disabled by default, that can be enabled with `--preflight=PermissionValidation=true`
```

##### Review Checklist:

- [x] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [x] Relevant tests are added or updated
- [x] Relevant docs in this repo added or updated
- [ ] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [x] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kapp/cmd/core/deps_factory.go
Comment (stevekuznetsov) [inline pkg/kapp/cmd/core/deps_factory.go]: Generally you want to use the deferred mapper so you don't send 300 requests to the server on startup.

File: pkg/kapp/permissions/validator.go
Comment (stevekuznetsov) [inline pkg/kapp/permissions/validator.go]: I don't think there's any way to handle the set of permissions in a Role without having loops for each of these dimensions - it's totally valid and reasonable to specify more than one verb, group, or resource (or all of them!) in a rule, as well as non-resource URLs, etc. That's why the OLMv0 bits we changed recently compute the hash of the rules and expect the user to have pre-created RBAC with the tool, so you can do a simple O(1) lookup.

Comment (joelanford) [reply pkg/kapp/permissions/validator.go]: I'd propose that we use `SelfSubjectRulesReview` here instead. I'm working on a kubectl plugin that might help or inspire: https://github.com/joelanford/kubectl-make_rbac

Comment (everettraven) [reply pkg/kapp/permissions/validator.go]: >I don't think there's any way to handle the set of permissions in a Role without having loops for each of these dimensions - it's totally valid and reasonable to specify more than one verb, group, or resource (or all of them!) in a rule, as well as non-resource URLs, etc.

Yep, due to the limitation of the `SelfSubjectAccessReview` requiring a single value reference for verb, group, and resource I used a helper that essentially iterates all these values and returns a slice of all the policies broken into a policy where only a single group, resource, and verb value are provided. 


> I'd propose that we use `SelfSubjectRulesReview` here instead. I'm working on a kubectl plugin that might help or inspire: https://github.com/joelanford/kubectl-make_rbac

This is also a direction I'm open to looking into. I've recently made some changes locally that will allow for an easier transition to whatever underlying implementation we feel is better suited for the use case here. Since this is still an early draft and a WIP, I'm going to focus on finishing up the plumbing before making any changes in the "how" the permissions are validated.

File: pkg/kapp/cmd/core/deps_factory.go
Comment (everettraven) [reply pkg/kapp/cmd/core/deps_factory.go]: Deferred mapper used in https://github.com/carvel-dev/kapp/pull/887/commits/feea616b47b08fef28ad4f0fbd565c600bb37575

File: pkg/kapp/cmd/appgroup/deploy.go
Comment (praveenrewar) [inline pkg/kapp/cmd/appgroup/deploy.go]: We do try to keep `kapp app-group deploy` and `kapp deploy` options in sync, so we should try to add the preflight checks here as well.

File: pkg/kapp/preflight/registry.go
Comment (praveenrewar) [inline pkg/kapp/preflight/registry.go]: I am wondering if having a map of all the preflight checks might become a little bit confusing in the help section 🤔
(I am a little bit inclined towards `--validate-options` as that seems more intuitive)

File: pkg/kapp/permissions/preflight.go
Comment (praveenrewar) [inline pkg/kapp/permissions/preflight.go]: As of now we don't use patch, but we do plan on using patch to add the last applied annotation, until then I think it's fine to skip patch.

File: pkg/kapp/cmd/app/deploy.go
Comment (praveenrewar) [inline pkg/kapp/cmd/app/deploy.go]: We can probably move this after we ask for a confirmation to deploy (`o.ui.AskForConfirmation`).

**Review by praveenrewar — COMMENTED** [review]
Looks pretty good, thank you so much for putting in the effort @everettraven ❤️

I do want to double check my understanding on the different validations. We are trying to ensure that:
- User has permissions to create the resources that they want to create.
- For rbac resources, they should only be able to add the permissions that they currently have? 

I didn't think of the second use case before, really glad that you have covered this as well 🙇🏻

File: pkg/kapp/cmd/app/deploy.go
Comment (100mik) [reply pkg/kapp/cmd/app/deploy.go]: Would we not need a permissions check before we diff? Or do we want to consider permissions needed for the apply?

Comment (praveenrewar) [reply pkg/kapp/cmd/app/deploy.go]: Before apply, we just need `get` permission, if we don't have that, we will fail very early.

Comment (100mik) [reply pkg/kapp/cmd/app/deploy.go]: I think it would be helpful to do it earlier, especially in case someone is trying put together some roles. Having this as a way of debugging roles seems to be pretty cool !

Comment (everettraven) [reply pkg/kapp/cmd/app/deploy.go]: Ah, moved it after confirmation before seeing this. I'll move it back :)

File: pkg/kapp/permissions/validator.go
Comment (everettraven) [reply pkg/kapp/permissions/validator.go]: Based on the discussion from the community meeting, the maintainers would prefer to get this in with SSAR and later investigate updating to use SSRR

File: pkg/kapp/preflight/registry.go
Comment (everettraven) [reply pkg/kapp/preflight/registry.go]: Maybe? This tries to keep it cleaner by only doing a slice of the check names in the usage method. If there is concern with include the list of defaults I can remove it. I personally err on the side of more information rather than less though. IMO it should be self documenting and the current implementation should adhere to that in as succinct a way as possible.

File: pkg/kapp/cmd/appgroup/deploy.go
Comment (everettraven) [reply pkg/kapp/cmd/appgroup/deploy.go]: Should be updated now

File: pkg/kapp/preflight/registry.go
Comment (praveenrewar) [reply pkg/kapp/preflight/registry.go]: My main concern was that the help section for preflight flag would grow to be very long as we add more preflight checks, but it looks fine as of now, so we can think about it later.

File: pkg/kapp/cmd/app/deploy.go
Comment (praveenrewar) [reply pkg/kapp/cmd/app/deploy.go]: My main reason for having it down was to ensure that the `diff` flags work even if we don't have enough permissions. For example, if I am using `--diff-run` along with `--preflight` I think that kapp should exit gracefully after showing the diff (even if we don't have sufficient permissions to make the changes) because we are _not_ making any change.

Comment (everettraven) [reply pkg/kapp/cmd/app/deploy.go]: Ah, that makes sense. Thanks @praveenrewar for the extra context! Does it make sense to move this after the diff is shown but before the confirmation step? I'm not really opinionated here because I'm not as well versed in the user flow here as you folks so I'm happy to move it wherever makes the most sense.

Comment (everettraven) [reply pkg/kapp/cmd/app/deploy.go]: Fixed linter failures and moved the preflight checks execution to right before the confirmation prompt in the latest push

Comment (praveenrewar) [reply pkg/kapp/cmd/app/deploy.go]: > Does it make sense to move this after the diff is shown but before the confirmation step?

Hmm, I was thinking that a user could still say "no" during the confirmation, but it wouldn’t really be a _preflight_ check if we do it after confirmation, so I think it makes sense.

File: test/e2e/preflight_permission_validation_escalation_test.go
Comment (praveenrewar) [inline test/e2e/preflight_permission_validation_escalation_test.go]: Could you update these comments. They are saying "missing" for all the tests cases.

Comment (praveenrewar) [reply test/e2e/preflight_permission_validation_escalation_test.go]: Should we also add a failing test case for escalation?

Comment (everettraven) [reply test/e2e/preflight_permission_validation_escalation_test.go]: Ah great catch on the comments. I meant to go back and update them. Must've gotten distracted and forgot 😁.

I'll go back and update the comments and add an e2e test specifically for failing privilege escalation

Comment (everettraven) [reply test/e2e/preflight_permission_validation_escalation_test.go]: In the latest push I:
- Removed the comments from all the tests as they aren't really necessary and I had originally used them as a check list for what I needed to implement
- Added an e2e test for failing privilege escalation

**Comment by everettraven** [comment]
@praveenrewar I've updated the e2e tests to fix the CI failure. Checked that it passed locally with a fresh cluster this time around.

**Review by praveenrewar — APPROVED** [review]
LGTM! Thank you so much for working on this!!
@100mik I am keeping it open in case you are still taking a look.

**Review by 100mik — APPROVED** [review]
LGTM! Thanks for the contribution ❤️ 

 @everettraven we will need some documentation changes to carvel-dev/carvel around the `--preflight` flag. I do not think it should block this PR! However, we should make sure we have these in place before we push out a release.

**Comment by 100mik** [comment]
I created a separate issue to track the required documentation changes 🙌🏼

**Pull request merged**
